src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs

```diff
@@ -358,8 +392,47 @@ private Task InvokeStreamCore(string methodName, InvocationRequest irq, object[]
             }
         }
 
+        private bool ProcessHandshakeResponse(ref ReadOnlyMemory<byte> data)
+        {
+            HandshakeResponseMessage message;
+
+            try
+            {
+                // read first message out of the incoming data
+                if (!TextMessageParser.TryParseMessage(ref data, out var payload))
+                {
+                    throw new InvalidDataException("Unable to parse payload as a handshake response message.");
+                }
+
+                message = HandshakeProtocol.ParseResponseMessage(payload.Span);
+            }
+            catch (Exception ex)
+            {
+                // shutdown if we're unable to read handshake
+                Log.ErrorReceivingHandshakeResponse(_logger, ex);
+                Shutdown(ex);
```

Comment:
What state are we in when we call this from the OnDataReceived callback? Does this fire the closed event inline?

Reply:
We should consider `Task.Run`ing the Closed event or something, but I don't think that's a problem isolated to this PR.

Reply:
Are we usually running user code here?